### PR TITLE
fix(deploy): guided deploy

### PR DESCRIPTION
**Pull request by sriram-mv**
- if attempting to deploy without build, sam cli does not crash
- stricter validation on `--image-repository` to always be an ECR URI
- if an ECR image repository is not provided not provided on guided, SAM
CLI does not crash.

#### Which issue(s) does this change fix?

* https://github.com/aws/aws-sam-cli/issues/2437

#### Why is this change necessary?
* Prevent SAM CLI crashes.

#### How does it address the issue?
* Add type validation to `--image-repository` option type.
#### What side effects does this change have?
* image repository can only be valid ECR URI
#### Checklist

- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [x] Write/update functional tests
- [x] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/package/regexpr.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+"""
+Regular Expressions for Resources.
+"""
+ECR_URL = "(^[a-zA-Z0-9][a-zA-Z0-9-_]*).dkr.ecr.([a-zA-Z0-9][a-zA-Z0-9-_]*).amazonaws.com(.cn)?/.*"
```
Comment (mgrandis) [inline samcli/lib/package/regexpr.py:4]: Should the dots be escaped like this "\."? Otherwise you are matching any char.

Comment (sriram-mv) [reply samcli/lib/package/regexpr.py:4]: That makes sense!
```
"(^[a-zA-Z0-9][a-zA-Z0-9-_]*)\.dkr\.ecr\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?\/.*"
```

is what i tested now and that seems to be more exact.

File: samcli/commands/deploy/guided_context.py [hunk 2]
```diff
@@ -270,13 +271,18 @@ def prompt_image_repository(self, parameter_overrides):
                 type=click.STRING,
                 default=self.image_repository if self.image_repository else "",
             )
+            if not image_repository or not is_ecr_url(image_repository):
```
Comment (mgrandis) [inline samcli/commands/deploy/guided_context.py:274]: I believe the `not image_repository` check is already done inside `is_ecr_url`.

File: tests/unit/cli/test_types.py [hunk 3]
```diff
@@ -285,3 +287,32 @@ def test_must_fail_on_invalid_format(self, input):
...
+
+class TestImageRepositoryType(TestCase):
+    def setUp(self):
+        self.param_type = ImageRepositoryType()
+        self.mock_param = Mock(opts=["--image-repository"])
+
+    @parameterized.expand(
+        [
+            # Just a string
+            ("some string"),
```
Comment (mgrandis) [inline tests/unit/cli/test_types.py:300]: You could maybe add more failing tests, with strings missing key parts of the pattern, like `123456789012.us-east-1.amazonaws.com/test1`

File: tests/unit/cli/test_types.py [hunk 4]
```diff
@@ -285,3 +287,32 @@ def test_must_fail_on_invalid_format(self, input):
...
+    def test_must_fail_on_invalid_format(self, input):
+        self.param_type.fail = Mock()
+        with self.assertRaises(BadParameter):
+            self.param_type.convert(input, self.mock_param, Mock())
+
+    @parameterized.expand(
+        [
+            (
+                "123456789012.dkr.ecr.us-east-1.amazonaws.com/test1",
+                "123456789012.dkr.ecr.us-east-1.amazonaws.com/test1",
```
Comment (mgrandis) [inline tests/unit/cli/test_types.py:312]: you can add a test for a `.cn` url: `123456789012.dkr.ecr.us-east-1.amazonaws.com.cn/test1`

File: samcli/commands/deploy/guided_context.py [hunk 5]
```diff
@@ -270,13 +271,18 @@ def prompt_image_repository(self, parameter_overrides):
...
+
             for _, function_prop in self.transformed_resources.functions.items():
                 if function_prop.packagetype == IMAGE:
                     image = function_prop.imageuri
                     try:
                         tag = tag_translation(image)
                     except NonLocalImageException:
                         pass
+                    except NoImageFoundException as ex:
+                        raise GuidedDeployFailedError("No images found to deploy, try sam build") from ex
```
Comment (mgrandis) [inline samcli/commands/deploy/guided_context.py:285]: I would add `running`: `try running sam build`

**Review by mgrandis — APPROVED** [review]

File: samcli/lib/package/regexpr.py [hunk 6]
```diff
@@ -0,0 +1,4 @@
+"""
+Regular Expressions for Resources.
+"""
+ECR_URL = r"(^[a-zA-Z0-9][a-zA-Z0-9-_]*)\.dkr\.ecr\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?\/.*"
```
Comment (hoffa) [inline samcli/lib/package/regexpr.py:4]: Just curious, is this documented anywhere? Or is it possible this will break some day?

**Review by hoffa — APPROVED** [review]

File: samcli/lib/package/regexpr.py (hunk 6 above)
Comment (sriram-mv) [reply samcli/lib/package/regexpr.py:4]: what we could document here is the docs section based on which the regex is derived.

https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_Repository.html does showcase the pattern for the repository name but not the entire regex.

What we can do is add a comment to link this above and also change the regex to allow for checking on the repository name as well.

**Pull request merged**